README.md
# BoxGrinder Brightbox Plugins

This archive integrates the Boxgrinder project [boxgrinder.org](http://www.boxgrinder.org) with the [Brightbox cloud CLI tools] (http://github.com/brightbox/brightbox-cli)

### Supported OSes

Brightbox supports all the OSs that Boxgrinder can generate.

### Requirements

* A RPM based build machine - use the current public Boxgrinder appliance registered on the Brightbox cloud
* This plugin gem installed.

### Installing

* `sudo gem install brightbox-boxgrinder-plugins` to install Brightbox CLI tools and the plugins onto the machine.
* `sudo brightbox-config client_add cli-xxxxx secretdetails` to add the api client details that allow access to the Brightbox cloud api.

### Usage

Create an appliance configuration file, then call the plugins as in the following example.

    boxgrinder-build --plugins brightbox-boxgrinder-plugins sl-basic.appl -p bbcloud -d bbcloud

To build an i386 version use:

    setarch i386 boxgrinder-build --plugins brightbox-boxgrinder-plugins sl-basic.appl -p bbcloud -d bbcloud

Obviously you are not limited to Scientific Linux and can create Fedora, RHEL or CentOS images too.



CHANGELOG
v0.0.6

* Really split FTP password properly

v0.0.5

* Split FTP reset password string properly

v0.0.4

* Remove sl plugin incorporated upstream
* Bring URLs and documentation up to date
* Upload images in Qcow compressed format

v0.0.3

* Switch dependencies to brightbox-cli

v0.0.2

* Ensure that the include file is the correct name

v0.0.1

* Inital Release


lib/os/sl-plugin.rb
#  Brightbox - Boxgrinder Scientific Linux OS Plugin
#  Copyright (c) 2011, Brightbox Systems
#  Author: Neil Wilson
#
# This is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation; either version 3 of
# the License, or (at your option) any later version.
#
# This software is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this software; if not, write to the Free
# Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
# 02110-1301 USA, or see the FSF site: http://www.fsf.org.

require 'boxgrinder-build/plugins/os/rhel/rhel-plugin'

module BoxGrinder
  class ScientificLinuxPlugin < RHELPlugin
   
    SL_REPOS = {
     "5" => {
       "base" => {
         "baseurl" => "http://ftp.scientificlinux.org/linux/scientific/#OS_VERSION#x/#BASE_ARCH#/SL/"
        },
       "updates" => {
         "baseurl" => "http://ftp.scientificlinux.org/linux/scientific/#OS_VERSION#x/#BASE_ARCH#/updates/security/"
        }
      },
     "6" => {
       "base" => {
         "baseurl" => "http://ftp.scientificlinux.org/linux/scientific/#OS_VERSION#x/#BASE_ARCH#/os/"
        },
       "updates" => {
         "baseurl" => "http://ftp.scientificlinux.org/linux/scientific/#OS_VERSION#x/#BASE_ARCH#/updates/security/"
        }
      }
    }


    def after_init
      super
      register_supported_os('sl', ['5', '6'])
    end

    def execute(appliance_definition_file)
      build_rhel(appliance_definition_file, SL_REPOS)
    end

  end
end

plugin :class => BoxGrinder::ScientificLinuxPlugin, :type => :os, :name => :sl, :full_name  => "Scientific Linux", :versions   => ["5", "6"]


lib/brightbox-boxgrinder-plugins.rb
#  Brightbox - Boxgrinder Plugins for Brightbox Cloud
#  Copyright (c) 2011, Brightbox Systems
#  Author: Neil Wilson
#
# This is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation; either version 3 of
# the License, or (at your option) any later version.
#
# This software is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this software; if not, write to the Free
# Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
# 02110-1301 USA, or see the FSF site: http://www.fsf.org.

require 'delivery/bbcloud-delivery-plugin'
require 'platform/bbcloud-platform-plugin'


lib/platform/bbcloud-platform-plugin.rb
#  Brightbox - Boxgrinder Platform Plugin for Brightbox Cloud
#  Copyright (c) 2011, Brightbox Systems
#  Author: Neil Wilson
#
# This is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation; either version 3 of
# the License, or (at your option) any later version.
#
# This software is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this software; if not, write to the Free
# Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
# 02110-1301 USA, or see the FSF site: http://www.fsf.org.

require 'boxgrinder-build/plugins/base-plugin'
require 'tempfile'

module BoxGrinder
  class BbcloudPlatformPlugin < BasePlugin
    plugin :type => :platform, :name => :bbcloud, :full_name => "Brightbox Cloud"

    def after_init
      register_deliverable(:disk => "#{deliverable_name}.qcow2")
      set_default_config_value('username', default_user)
      @appliance_config.packages |= %w{cloud-init} if has_cloud_init?
    end

    def execute
      @log.info "Converting #{@appliance_config.name} appliance image to use Brightbox metadata services"

      @log.debug "Using qemu-img to convert the image to qcow2 format..."
      @image_helper.convert_disk(@previous_deliverables.disk, :qcow2, @deliverables.disk)
      @log.debug "Conversion done."

      @log.debug "Adding metadata customisations"
      @image_helper.customize([@deliverables.disk], :automount => true) do |guestfs, guestfs_helper|
        add_default_user(guestfs)
	disable_root_password(guestfs_helper)
	set_default_timezone(guestfs_helper)
        change_ssh_configuration(guestfs_helper)
	if has_cloud_init?
	  customise_cloud_init(guestfs_helper)
	else
	  update_rc_local(guestfs)
	end
        execute_post(guestfs_helper)
      end
      @log.debug "Added customisations"
    end

    def has_cloud_init?
      @appliance_config.os.name == 'fedora' and @appliance_config.os.version >= '16'
    end

    def customise_cloud_init(guestfs_helper)
      guestfs_helper.sh("sed -i 's/^user: ec2-user$/user: #{@plugin_config['username']}/' /etc/cloud/cloud.cfg")
    end

    def execute_post(guestfs_helper)
      if @appliance_config.post['bbcloud']
        @appliance_config.post['bbcloud'].each do |cmd|
          guestfs_helper.sh(cmd, :arch => @appliance_config.hardware.arch)
        end
        @log.debug "Post commands from appliance definition file executed."
      else
        @log.debug "No Post commands specified, skipping."
      end
    end

    def default_user
      "brightbox"
    end

    def deliverable_name
      "#{@appliance_config.name}-#{@appliance_config.version}.#{@appliance_config.release}-#{@appliance_config.os.name}-#{@appliance_config.os.version}"
    end

    def add_default_user(guestfs)
      @log.debug "Adding #{@plugin_config['username']} user..."
      unless guestfs.fgrep(@plugin_config['username'], "/etc/passwd").empty?
        @log.debug("#{@plugin_config['username']} already exists, skipping.")
        return
      end
      guestfs.sh("useradd #{@plugin_config['username']}")
      guestfs.sh("echo -e '#{@plugin_config['username']}\tALL=(ALL)\tNOPASSWD: ALL' >> /etc/sudoers")
      @log.debug "User #{@plugin_config['username']} added."
    end

    def disable_root_password(guestfs_helper)
      @log.debug "Disabling root password"
      guestfs_helper.sh("usermod -L root")
      @log.info "root password disabled - use sudo from #{@plugin_config['username']} account"
    end

    def update_rc_local(guestfs)
      @log.debug "Updating '/etc/rc.d/rc.local' file..."
      rc_local = Tempfile.new('rc_local')

      if guestfs.exists("/etc/rc.d/rc.local") == 1
        # We're appending
        rc_local << guestfs.read_file("/etc/rc.d/rc.local")
      else
        # We're creating new file
        rc_local << "#!/bin/bash\n\n"
      end

      rc_local << File.read("#{File.dirname(__FILE__)}/src/rc_local")
      rc_local.flush

      guestfs.upload(rc_local.path, "/etc/rc.d/rc.local")

      rc_local.close
    end

    def change_ssh_configuration(guestfs_helper)
      guestfs_helper.augeas do
        # disable password authentication
        set("/etc/ssh/sshd_config", "PasswordAuthentication", "no")

        # disable root login
        set("/etc/ssh/sshd_config", "PermitRootLogin", "no")

	# Switch off GSS Authentication
        set("/etc/ssh/sshd_config", "GSSAPIAuthentication", "no")
      end
    end

    def set_default_timezone(guestfs_helper)
      guestfs_helper.augeas do
        set("/etc/sysconfig/clock", "UTC", "True")
        set("/etc/sysconfig/clock", "ZONE", "Etc/UTC")
      end
      guestfs_helper.guestfs.cp("/usr/share/zoneinfo/UTC", "/etc/localtime")
    end

  end
end



lib/delivery/bbcloud-delivery-plugin.rb
#  Brightbox - Boxgrinder Delivery Plugin for Brightbox Cloud
#  Copyright (c) 2011, Brightbox Systems
#  Author: Neil Wilson
#
# This is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as
# published by the Free Software Foundation; either version 3 of
# the License, or (at your option) any later version.
#
# This software is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this software; if not, write to the Free
# Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
# 02110-1301 USA, or see the FSF site: http://www.fsf.org.

require 'boxgrinder-build/plugins/base-plugin'

module BoxGrinder
  class BbcloudDeliveryPlugin < BasePlugin

    plugin :type => :delivery, :name => :bbcloud, :full_name => "Brightbox Cloud Image Registration Service"
    def after_init
      set_default_config_value('description', default_description)
      set_default_config_value('name', default_name)
      set_default_config_value('username', default_username)
      set_default_config_value('mode', default_mode)
      set_default_config_value('public', default_public)
      validate_mode
      validate_public
    end
   
    def execute( type = :bbcloud )
      @log.info "Adding '#{@plugin_config['name']}' appliance to Brightbox Cloud..." 
      @log.info "Using Brightbox account id #{account}"
      upload
      register_image
    end

    def upload
      @log.info "Uploading to #{ftp_hash['library_ftp_host']} with secure FTP"
      if system curl_command
        @log.info "Appliance #{@plugin_config['name']} uploaded."
      else
        raise "An error occurred while uploading files."
      end
    end

    def register_image
      @log.info "Registering appliance as #{image_id} under account #{account} with the name '#{@plugin_config['name']}'"
      @log.info "Run 'brightbox-images show #{image_id}' to check registration progress"
    end

    def image_id
      @image_id ||=
	if @exec_helper.execute(register_image_command) =~ /img-\w{5}/
	  Regexp.last_match[0]
	else
	  raise "Failed to obtain an image id from the registration command"
	end
    end

    def account
      @account ||= @exec_helper.execute("brightbox-accounts -s list 2>/dev/null").split[0]
    rescue RuntimeError => e
      @log.error e.message
      raise PluginValidationError, "Make sure the that brightbox cloud API tools are installed. Use 'brightbox-config client_add' to add the api client details for your account." 
    end

    def disk_image
      @previous_deliverables[:disk]
    end

    def ftp_hash
      @ftp_hash ||= Hash[*(@exec_helper.execute("brightbox-accounts -s reset_ftp_password #{account} 2>/dev/null").split(/[\t\n]/))]
    end

    def target_name
      File.basename(disk_image)
    end

    def curl_command
      "curl #{if ENV['INSECURE'] then "-k" end} -# -u #{ftp_hash['library_ftp_user']}:#{ftp_hash['library_ftp_password']} --ftp-ssl-control -T #{disk_image} ftp://#{ftp_hash['library_ftp_host']}/incoming/#{target_name}"
    end
    
    def default_name
      @appliance_config.name
    end

    def default_description
      @appliance_config.summary
    end

    def default_mode
      'virtio'
    end

    def default_username
      'brightbox'
    end

    def default_public
      'false'
    end

    def validate_mode
      case @plugin_config['mode']
      when 'compatibility', 'virtio'
      else
        raise PluginValidationError, "Valid values for #{@plugin_info[:full_name]} plugin 'mode' option are 'compatibility' or 'virtio'"
      end
    end

    def validate_public
      case @plugin_config['public']
      when 'true', 'false'
      else
        raise PluginValidationError, "Valid values for #{@plugin_info[:full_name]} plugin 'public' option are 'true' or 'false'"
      end
    end

    def register_image_command
      "brightbox-images register -a #{@appliance_config.hardware.arch} -s #{target_name} -n '#{@plugin_config['name']}' -d '#{@plugin_config['description']}' -u '#{@plugin_config['username']}' -m '#{@plugin_config['mode']}' -p '#{@plugin_config['public']}'"
    end

  end
end

